File: dfs/dfs_combination.py
#-*-coding:utf-8-*-
def combination(n: int, k: int):
    results = []
    def dfs(elements, start: int, k: int):
        if k == 0:
            results.append(elements[:])
            return
        # 자신 이전의 모든 값을 고정하여 재귀 호출
        for i in range(start, n+1):
            elements.append(i)
            dfs(elements, i+1, k-1)
            elements.pop()
    dfs([], 1, k)
    return results


def combination_list(nums:list, k:int):
    results = []
    print("combination_list()")

    def dfs(elements, start:int, k:int):
        if k == 0:
            results.append(elements[:])
            return
        # elif k < 0:
        #     return
        
        for i in range(start, len(nums)):
            elements.append(nums[i])
            dfs(elements, i+1, k-1)
            elements.pop()
    
    dfs([], 0, k)
    return results

def comb_l2(nums:list[int], k:list):
    results = []

    def dfs(elements, start:int, k:int):
        if k == 0:
            results.append(elements[:])
            return
        
        for i in range(start, len(nums)):
            elements.append(nums[i])
            dfs(elements, i+1, k-1)
            elements.pop()
        

if __name__ == "__main__":
    reusult = combination(4, 2)
    print(reusult)
    nums = [1, 2, 3, 4]
    result_list = combination_list(nums, 2)
    print(result_list)

File: dfs/dfs_combination.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> results;

void dfs_combination(vector<int>& elem, vector<int>& nums, int start, int k) {
	if (k == 0) {
		results.push_back(elem);
		return;
	}
	for (int i = start; i < nums.size(); i++) {
		elem.push_back(nums[i]);
		dfs_combination(elem, nums, i + 1, k - 1);
		elem.pop_back();
	}
}

void combination(vector<int>& nums, int k) {
	vector<int> elems;
	dfs_combination(elems, nums, 0, k);
}

int main()
{
	//cout << "main(): " << endl;

	vector<int> nums = { 1, 2, 3, 4, 5 };
	combination(nums, 3);

	cout << "results: " << endl;
	cout << "{" << endl;
	for (auto v : results) {
		cout << "  {";
		for (auto d : v) {
			cout << d << ", ";
		}
		cout << "}" << endl;
	}
	cout << "}" << endl;

	return 0;
}
